tests/integration/test_node_port.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+
+import psycopg2
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+
+from . import markers
+from .helpers.ha_helpers import (
+    start_continuous_writes,
+    stop_continuous_writes,
+)
+from .helpers.helpers import (
+    CHARM_SERIES,
+    CLIENT_APP_NAME,
+    PGB,
+    PGB_METADATA,
+    POSTGRESQL_APP_NAME,
+    app_name,
+    get_endpoint_info,
+    get_juju_secret,
+    get_unit_info,
+)
+from .juju_ import juju_major_version
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+DATA_INTEGRATOR = "data-integrator"
+MODEL_CONFIG = {"logging-config": "<root>=INFO;unit=DEBUG"}
+
+
+if juju_major_version < 3:
+    TLS_CERTIFICATES_APP_NAME = "tls-certificates-operator"
+    TLS_CHANNEL = "legacy/stable"
+    TLS_CONFIG = {"generate-self-signed-certificates": "true", "ca-common-name": "Test CA"}
+else:
+    TLS_CERTIFICATES_APP_NAME = "self-signed-certificates"
+    TLS_CHANNEL = "latest/stable"
+    TLS_CONFIG = {"ca-common-name": "Test CA"}
+DATABASE_UNITS = 3
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest, pgb_charm):
+    """Test the deployment of the charm."""
+    # Build and deploy applications
+    await ops_test.model.set_config(MODEL_CONFIG)
+    wait_for_apps = False
+
+    if not await app_name(ops_test):
+        wait_for_apps = True
+        async with ops_test.fast_forward():
+            await ops_test.model.deploy(
+                pgb_charm,
+                resources={
+                    "pgbouncer-image": PGB_METADATA["resources"]["pgbouncer-image"][
+                        "upstream-source"
+                    ]
+                },
+                application_name=PGB,
+                num_units=1,
```

Comment:
As discussed in private, IMHO: we must test 2-3 units here. As in production we have 2+ PGB units for HA. 

I believe each client app have to deploy independent mysql-router (for VMs for sure... but I would keep this requirements for K8s as well). Example: pgb normally should be deployen in app mode, so wordpress in model A will have pgb in the same model, but data-integrator will in the DB model OR separate (but outside of model A => different PGB in the real deployment).

Sure, it theory the same PGB should work for 2+ apps... but create a separate bugreport if it is broken here... and we should re-check legacy interface behavior here (app1:psql + app2:postgresql_client => BOOM?).